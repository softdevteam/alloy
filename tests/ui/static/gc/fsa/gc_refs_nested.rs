#![feature(gc)]
#![feature(negative_impls)]
#![allow(dead_code)]
#![allow(unused_variables)]
include!{"./auxiliary/types.rs"}

impl Drop for HasNestedGc {
    fn drop(&mut self) {
        use_val(self.a); // should fail
        use_val(self.b); // should pass

        // Project through to the nested `HasGc` struct.
        use_val(self.c.a); // should fail
        use_val(self.c.b); // should pass

        let a = self.a; // should fail
        let b = self.b; // should pass

        let ca = self.a; // should fail
        let cb = self.b; // should pass

        // should pass, as not a field projection
        let d = &1;
        use_val(d);
    }
}

fn main() {
    Gc::new(HasNestedGc::default());
    //~^     ERROR: The drop method for `HasNestedGc` cannot be safely finalized.
    //~|     ERROR: The drop method for `HasNestedGc` cannot be safely finalized.
    //~|     ERROR: The drop method for `HasNestedGc` cannot be safely finalized.
    //~|     ERROR: The drop method for `HasNestedGc` cannot be safely finalized.
}
