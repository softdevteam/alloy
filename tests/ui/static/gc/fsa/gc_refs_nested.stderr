error: The drop method for `HasNestedGc` cannot be safely finalized.
  --> $DIR/gc_refs_nested.rs:29:13
   |
LL |         use_val(self.a); // should fail
   |                 ------
   |                 |
   |                 a finalizer cannot safely dereference this `Gc<u64>`
   |                 from a drop method because it might have already been finalized.
...
LL |     Gc::new(HasNestedGc::default());
   |     --------^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasNestedGc>` here.

error: The drop method for `HasNestedGc` cannot be safely finalized.
  --> $DIR/gc_refs_nested.rs:29:13
   |
LL |         use_val(self.c.a); // should fail
   |                 --------
   |                 |
   |                 a finalizer cannot safely use this `HasGc`
   |                 from a drop method because it does not implement `FinalizerSafe`.
...
LL |     Gc::new(HasNestedGc::default());
   |     --------^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasNestedGc>` here.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values whose types implement `FinalizerSafe`.

error: The drop method for `HasNestedGc` cannot be safely finalized.
  --> $DIR/gc_refs_nested.rs:29:13
   |
LL |         use_val(self.c.b); // should pass
   |                 --------
   |                 |
   |                 a finalizer cannot safely use this `HasGc`
   |                 from a drop method because it does not implement `FinalizerSafe`.
...
LL |     Gc::new(HasNestedGc::default());
   |     --------^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasNestedGc>` here.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values whose types implement `FinalizerSafe`.

error: The drop method for `HasNestedGc` cannot be safely finalized.
  --> $DIR/gc_refs_nested.rs:29:13
   |
LL |         let a = self.a; // should fail
   |                 ------
   |                 |
   |                 a finalizer cannot safely dereference this `Gc<u64>`
   |                 from a drop method because it might have already been finalized.
...
LL |     Gc::new(HasNestedGc::default());
   |     --------^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasNestedGc>` here.

error: The drop method for `HasNestedGc` cannot be safely finalized.
  --> $DIR/gc_refs_nested.rs:29:13
   |
LL |         let ca = self.a; // should fail
   |                  ------
   |                  |
   |                  a finalizer cannot safely dereference this `Gc<u64>`
   |                  from a drop method because it might have already been finalized.
...
LL |     Gc::new(HasNestedGc::default());
   |     --------^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<HasNestedGc>` here.

error: aborting due to 5 previous errors

